Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,586 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+
+import Darwin
+import CNIOLinux
+import CoreFoundation
+
+
+/// `Error` that may be thrown if we fail to create a `IPAddress`
+public enum IPAddressError: Error {
+    /// Given string input is not supported IP Address
+    case failedToParseIPString(String)
+    /// Given string input is not supported IP Address
+    case bytesArrayHasWrongLength(Int)
+}
+
+public typealias IPv4BytesTuple = (UInt8, UInt8, UInt8, UInt8)
+public typealias IPv6BytesTuple = (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8)
+
+/// Represent the bytes for an `IPv4Address`.
+public struct IPv4Bytes {
+    public var bytes: IPv4BytesTuple
+    
+    public init(_ bytes: IPv4BytesTuple) {
+        self.bytes = bytes
+    }
+}
+
+/// Represent the bytes for an `IPv6Address.
+public struct IPv6Bytes {
+    public var bytes: IPv6BytesTuple
+    
+    public init(_ bytes: IPv6BytesTuple) {
+        self.bytes = bytes
+    }
+}
+
+/// Represent a single `IPAddress`.
+public enum IPAddress: CustomStringConvertible {
+    /// A single IPv4 address for `IPAddress`.
+    public struct IPv4Address: CustomStringConvertible {
+        /// The bytes storing the address of the IPv4 address.
+        public var address: IPv4Bytes
+        
+        /// Get the `IPv4Address` as a string.
+        public var ipAddressString: String {
+            self.address.lazy.map({String($0)}).joined(separator: ".")
+        }
+        
+        /// A human-readable description of this `IPv4Address`. Mostly useful for logging.
+        public var description: String {
+            return "[IPv4]\(self.ipAddressString)"
+        }
+        
+        /// Get the libc address for an IPv4 address.
+        public var posix: in_addr {
+            get {
+                return in_addr.init(s_addr: UInt32(uint8Tuple: self.address.bytes))
+            }
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - address: Bytes that hold the IPv4 address.
+        public init(address: IPv4Bytes) {
+            self.address = address
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - packedBytes: Collection of UInt8 that holds the address.
+        @inlinable
+        public init<Bytes: Collection>(packedBytes bytes: Bytes) throws where Bytes.Element == UInt8 {
+            if bytes.count != 4 {
+                throw IPAddressError.bytesArrayHasWrongLength(bytes.count)
+            }
+            
+            var ipv4Bytes = IPv4Bytes((0,0,0,0))
+            
+            for (idx, elt) in bytes.prefix(4).enumerated() {
```

Comment:
We can remove `prefix(4)` here: it doesn't achieve anything with the check above.